--- eglibc-2.15/ports/sysdeps/unix/sysv/linux/mips/Makefile	2011-12-22 20:27:39.000000000 +0100
+++ eglibc-2.15-mips-no-n32-n64/ports/sysdeps/unix/sysv/linux/mips/Makefile	2012-02-24 16:23:23.380420249 +0100
@@ -7,7 +7,7 @@
 
 sysdep_headers += sys/cachectl.h sys/sysmips.h sys/tas.h
 
-syscall-list-variants := o32 n32 n64
+syscall-list-variants := o32
 syscall-list-includes := sgidefs.h
 syscall-list-o32-options := -mabi=32
 syscall-list-o32-condition := _MIPS_SIM == _MIPS_SIM_ABI32
