--- eglibc-2.15/ports/sysdeps/mips/Makefile	2012-01-04 17:35:09.000000000 +0100
+++ eglibc-2.15-mips-async-unwind/ports/sysdeps/mips/Makefile	2012-02-24 15:19:44.915827395 +0100
@@ -21,11 +21,11 @@
 endif
 
 ifeq ($(subdir),csu)
-CFLAGS-initfini.s += -fno-unwind-tables
+CFLAGS-initfini.s += -fno-asynchronous-unwind-tables -fno-unwind-tables 
 endif
 
 ifeq ($(subdir),nptl)
-CFLAGS-pt-initfini.s += -fno-unwind-tables
+CFLAGS-pt-initfini.s += -fno-asynchronous-unwind-tables -fno-unwind-tables
 endif
 
 ASFLAGS-.os += $(pic-ccflag)
