diff --git a/ports/sysdeps/aarch64/nptl/pthread_spin_lock.c b/ports/sysdeps/aarch64/nptl/pthread_spin_lock.c
index 490bd43..7f2505f 100644
--- a/ports/sysdeps/aarch64/nptl/pthread_spin_lock.c
+++ b/ports/sysdeps/aarch64/nptl/pthread_spin_lock.c
@@ -21,4 +21,4 @@
 /* We can't use the normal "#include <nptl/pthread_spin_lock.c>" because
    it will resolve to this very file.  Using "sysdeps/.." as reference to the
    top level directory does the job.  */
-#include <sysdeps/../nptl/pthread_spin_lock.c>
+#include <sysdeps/../../eglibc-2.19/nptl/pthread_spin_lock.c>
diff --git a/ports/sysdeps/arm/nptl/pthread_spin_lock.c b/ports/sysdeps/arm/nptl/pthread_spin_lock.c
index 7105c73..795db52 100644
--- a/ports/sysdeps/arm/nptl/pthread_spin_lock.c
+++ b/ports/sysdeps/arm/nptl/pthread_spin_lock.c
@@ -20,4 +20,4 @@
 /* We can't use the normal "#include <nptl/pthread_spin_lock.c>" because
    it will resolve to this very file.  Using "sysdeps/.." as reference to the
    top level directory does the job.  */
-#include <sysdeps/../nptl/pthread_spin_lock.c>
+#include <sysdeps/../../eglibc-2.19/nptl/pthread_spin_lock.c>
diff --git a/ports/sysdeps/mips/nptl/pthread_spin_lock.c b/ports/sysdeps/mips/nptl/pthread_spin_lock.c
index f3e718c..a6c4ca3 100644
--- a/ports/sysdeps/mips/nptl/pthread_spin_lock.c
+++ b/ports/sysdeps/mips/nptl/pthread_spin_lock.c
@@ -20,4 +20,5 @@
 /* We can't use the normal "#include <nptl/pthread_spin_lock.c>" because
    it will resolve to this very file.  Using "sysdeps/.." as reference to the
    top level directory does the job.  */
-#include <sysdeps/../nptl/pthread_spin_lock.c>
+#include <sysdeps/../../eglibc-2.19/nptl/pthread_spin_lock.c>
+
